Objects/regions.c

```diff
@@ -160,28 +195,76 @@ typedef struct _gc_runtime_state GCState;
 #define GC_PREV _PyGCHead_PREV
 #define FROM_GC(g) ((PyObject *)(((char *)(g))+sizeof(PyGC_Head)))
 
+static int
+is_immutable_region(Py_uintptr_t r)
+{
+    return r == _Py_IMMUTABLE;
+}
+
+static int
+is_default_region(Py_uintptr_t r)
+{
+    return r == _Py_DEFAULT_REGION;
+}
 
 /* A traversal callback for _Py_CheckRegionInvariant.
    - op is the target of the reference we are checking, and
    - parent is the source of the reference we are checking.
 */
 static int
-visit_invariant_check(PyObject *op, void *parent)
+visit_invariant_check(PyObject *tgt, void *parent)
 {
     PyObject *src_op = _PyObject_CAST(parent);
-    // Check Immutable only reaches immutable
-    if ((src_op->ob_region == _Py_IMMUTABLE)
-        && (op->ob_region != _Py_IMMUTABLE))
-        {
-            set_failed_edge(src_op, op, "Destination is not immutable");
-            return 0;
-        }
-    // TODO: More checks to go here as we add more region
-    // properties.
+    regionmetadata* src_region = (regionmetadata*) Py_GET_REGION(src_op);
+    regionmetadata* tgt_region = (regionmetadata*) Py_GET_REGION(tgt);
+     // Internal references are always allowed
+    if (src_region == tgt_region)
+        return 0;
+    // Anything is allowed to point to immutable
+    if (is_immutable_region((Py_uintptr_t)tgt_region))
+        return 0;
+    // Borrowed references are unrestricted
+    if (is_default_region((Py_uintptr_t)src_region))
+        return 0;
+    // Since tgt is not immutable, src also may not be as immutable may not point to mutable
+    if (is_immutable_region((Py_uintptr_t)src_region)) {
+        set_failed_edge(src_op, tgt, "Destination is not immutable");
+        return 0;
+    }
+
+    // Cross-region references must be to a bridge
+    if (!is_bridge_object(tgt)) {
+        set_failed_edge(src_op, tgt, "Destination is not in the same region");
+        return 0;
+    }
+    // Check if region is already added to captured list
+    if (tgt_region->next != NULL) {
```

Comment:
IDK if that would actually help, I find this more readable and since it's the only place I'd leave it for now, but I don't have a super strong opinion on it.

Reply:
I'm fine with that reasoning.